### PR TITLE
AAP-32143 Make the JT name uniqueness enforced at the database level

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
This was pretty clearly introduced with the introduction of named URLs.

You can see the reproducer in the test. It literally spawns subprocesses that makes JTs in parallel. I tuned it down to 5 processes, because I found that using any more won't create any more duplicates. Probably because the `requests` library limits us to 5 concurrent requests. Interestingly though, **all 5** manage to create a JT with the duplicated name. So at this time scale, there basically _is no_ unique name enforcement.

TODO:
 - add a migration to rename the duplicates... because what else can we do?
 - run tests
 - Renumber migration, related to https://github.com/ansible/awx/pull/15914

##### ISSUE TYPE
 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME
 - API

### CODE REVIEW

**Comment by AlanCoding** [comment]
Noting test failures
 - test_project_list_ordering_with_duplicate_names
 - awx/main/tests/functional/test_copy.py::test_job_template_copy
 - awx/main/tests/functional/task_management/test_scheduler.py::test_dependency_isolation
 - awx/main/tests/functional/migrations/test_inventory_source_migration.py::test_cloudforms_inventory_removal
 - awx/main/tests/functional/analytics/test_counts.py::test_inventory_counts

Where these tests just being lazy before? Or is there a real conflict of expectations? Might be a mix.

**Comment by AlanCoding** [comment]
All the tests are actually coming back looking okay at this point.

**Review by djyasin — APPROVED** [review]

File: awx/main/migrations/0203_template_name_constraint.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+logger = logging.getLogger(__name__)
+
+
+def _rename_duplicates(cls):
+    for organization_id in cls.objects.order_by().values_list('organization_id', flat=True).distinct():
+        duplicate_data = cls.objects.values('name').filter(organization_id=organization_id).annotate(count=Count('name')).order_by().filter(count__gt=1)
+        for data in duplicate_data:
+            name = data['name']
+            ct = 0
+            for ujt in cls.objects.filter(name=name):
```
Comment (fosterseth) [inline awx/main/migrations/0203_template_name_constraint.py]: we could considering only renaming the dups, not all of them. Maybe the most recent (by creation time) gets to keep the name, the rest get the suffix?

as a counter, maybe it's good to throw some "jt is missing" if automation is hitting these renamed JTs, so that the user can go back and see the renaming that was done and properly choose the JT they want.

File: awx/main/tests/functional/test_projects.py [hunk 2]
```diff
@@ -436,21 +436,21 @@ def test_project_list_ordering_by_name(get, order_by, expected_names, organizati
 
 @pytest.mark.parametrize('order_by', ('name', '-name'))
 @pytest.mark.django_db
-def test_project_list_ordering_with_duplicate_names(get, order_by, organization_factory):
+def test_project_list_ordering_with_duplicate_names(get, order_by, organization_factory, admin):
```
Comment (fosterseth) [inline awx/main/tests/functional/test_projects.py]: does this test even make sense anymore?

File: awx/main/tests/live/tests/api/test_uniqueness.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        print(f'Created user {user.username}')
+    return user
+
+
+@pytest.fixture
+def admin_auth(admin_for_test):
+    return HTTPBasicAuth(admin_for_test.username, 'for_test_123!')
+
+
+def test_jt_duplicate_name(admin_auth, demo_proj):
```
Comment (fosterseth) [inline awx/main/tests/live/tests/api/test_uniqueness.py:52]: I was expecting something simpler here

at django level
```
create_jt('bob')
create_jt('bob') -> assert db constraint error
```

at api level
```
api_create_jt('bob') > 200
api_create_jt('bob') > 400
```

am I missing some edge case that a simpler test approach wouldn't be capturing, that this live test does capture?

Comment (AlanCoding) [reply awx/main/tests/live/tests/api/test_uniqueness.py:52]: The issue is that we have a pre-check, and the bug can only be hit with a race condition. If you hit the race condition, you actually get a different error than when the pre-check happens. People wrote these sort of tests and assumed everything was working right, but it wasn't.

Django also has some pre-check logic, but our own logic is doing the SOFT_UNIQUE_TOGETHER thing. That doesn't work for parallel requests because of the read-write behavior of transactions. Process 1 is unaware that process 2 has created a JT of the same name because neither process has committed their changes. What this change does is cause 1 of the 2 to hit a database error on the commit. That is unthinkable to do with sqlite3, or within the pytest-django transactional constraints which prevents ever actually committing anything.

File: awx/main/tests/functional/test_projects.py (hunk 2 above)
Comment (fosterseth) [reply awx/main/tests/functional/test_projects.py]: nvm the test is still valid in ensuring stable return of project list. I think the test title could be updated to remove the "duplicate names" part

File: awx/main/migrations/0203_template_name_constraint.py (hunk 1 above)
Comment (AlanCoding) [reply awx/main/migrations/0203_template_name_constraint.py]: > as a counter, maybe it's good to throw some "jt is missing" if automation is hitting these renamed JTs

This is very complex communication-wise. Thinking more about it, realize that this would block upgrade being done by a system admin, due to what some joe schmo user did.

It might be a reasonable improvement to not rename the oldest template. I am leaning towards making that our 1 and only singular solution we offer.

File: awx/main/tests/functional/test_projects.py (hunk 2 above)
Comment (AlanCoding) [reply awx/main/tests/functional/test_projects.py]: This is a case where changing the test title wouldn't be completely honest. I undid some stuff that I shouldn't have. I'm going to push a commit that reverts part of this change, and puts it closer to the prior behavior. You can have duplicate names, just not in the same organization, so that's what we should be doing here.

File: awx/main/migrations/0203_template_name_constraint.py (hunk 1 above)
Comment (AlanCoding) [reply awx/main/migrations/0203_template_name_constraint.py]: Pushed a commit that implements this. Added some new testing to help get that worked out.

**Review by fosterseth — APPROVED** [review]

**Review by chrismeyersfsu — APPROVED** [review]

**Review by adrisala — APPROVED** [review]
